module/os/linux/zfs/vdev_disk.c

```diff
@@ -252,15 +270,29 @@ vdev_disk_open(vdev_t *v, uint64_t *psize, uint64_t *max_psize,
 					reread_part = B_TRUE;
 			}
 
+#ifdef HAVE_BLKDEV_PUT_HOLDER
+			blkdev_put(bdev, zfs_vdev_holder);
+#else
 			blkdev_put(bdev, mode | FMODE_EXCL);
+#endif
 		}
 
 		if (reread_part) {
+#ifdef HAVE_BLKDEV_GET_BY_PATH_4ARG
+			bdev = blkdev_get_by_path(disk_name,
```

Comment:
I think it'd be helpful for readability to create `vdev_blkdev_get_by_path()` and `blkdev_put()` wrappers.  Something like this (untested):

```c
static inline struct block_device *
vdev_blkdev_get_by_path(spa_mode_t mode, const char *path, void *holder,
    const struct blk_holder_ops *hops)
{       
#ifdef HAVE_BLKDEV_GET_BY_PATH_4ARGS
        return (blkdev_get_by_path(path,
            vdev_bdev_mode(mode) | BLK_OPEN_EXCL, holder, hops);
#else   
        return (blkdev_get_by_path(path,
            vdev_bdev_mode(mode) | FMODE_EXCL, holder);
#endif
}

static inline void
vdev_blkdev_put(spa_mode_t *spa, struct block_device *bdev, void *holder)
{
#ifdef HAVE_BLKDEV_PUT_HOLDER
        return (bdev, holder);
#else
        return (blkdev_put(bdev,  vdev_bdev_mode(mode) | FMODE_EXCL));
#endif
}
```